### PR TITLE
Add cache for LookupStateReference

### PR DESCRIPTION
**Pull request by earlbread**
This adds the LRU cache for LookupStateReference.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -105,6 +106,7 @@ To be released.
 [#709]: https://github.com/planetarium/libplanet/pull/709
 [#718]: https://github.com/planetarium/libplanet/pull/718
 [#719]: https://github.com/planetarium/libplanet/pull/719
+[#720]: https://github.com/planetarium/libplanet/pull/720
```
Comment (moreal) [inline CHANGES.md]:
```suggestion
[#722]: https://github.com/planetarium/libplanet/pull/722
```

File: Libplanet/Store/DefaultStore.cs [hunk 2]
```diff
@@ -564,6 +569,54 @@ HashDigest<SHA256> blockHash
...
+        {
+            if (lookupUntil is null)
+            {
+                throw new ArgumentNullException(nameof(lookupUntil));
+            }
+
+            if (_lastStateRefCaches.TryGetValue(chainId, out var stateRefCache)
+                && stateRefCache.TryGetValue(address, out var cache))
+            {
+                var cachedIndex = cache.Item2;
```
Comment (longfin) [inline Libplanet/Store/DefaultStore.cs]:
```suggestion
                long cachedIndex = cache.Item2;
```

File: Libplanet/Store/DefaultStore.cs [hunk 3]
```diff
@@ -564,6 +569,54 @@ HashDigest<SHA256> blockHash
...
+
+                if (cachedIndex <= lookupUntil.Index)
+                {
+                    return cache;
+                }
+            }
+
+            var stateRef =
```
Comment (longfin) [inline Libplanet/Store/DefaultStore.cs]:
```suggestion
            Tuple<HashDigest<SHA256>, long> stateRef =
```

File: Libplanet/Store/DefaultStore.cs [hunk 4]
```diff
@@ -619,6 +672,24 @@ HashDigest<SHA256> blockHash
...
             coll.EnsureIndex("AddressString");
             coll.EnsureIndex("BlockIndex");
+
+            if (!_lastStateRefCaches.ContainsKey(chainId))
+            {
+                _lastStateRefCaches[chainId] =
+                    new LruCache<Address, Tuple<HashDigest<SHA256>, long>>();
+            }
+
+            var stateRefCache = _lastStateRefCaches[chainId];
```
Comment (longfin) [inline Libplanet/Store/DefaultStore.cs]:
```suggestion
            LruCache<Address, Tuple<HashDigest<SHA256>, long>> stateRefCache = _lastStateRefCaches[chainId];
```

File: Libplanet/Store/DefaultStore.cs [hunk 5]
```diff
@@ -619,6 +672,24 @@ HashDigest<SHA256> blockHash
...
+
+            foreach (var address in addresses)
+            {
+                _logger.Debug($"Try to set cache {address}");
+                if (!stateRefCache.TryGetValue(address, out var cache) || cache.Item2 < blockIndex)
```
Comment (longfin) [inline Libplanet/Store/DefaultStore.cs]:
```suggestion
                if (!stateRefCache.TryGetValue(address, out Tuple<HashDigest<SHA256>, long> cache) || cache.Item2 < blockIndex)
```

File: Libplanet/Store/DefaultStore.cs [hunk 6]
```diff
@@ -564,6 +569,54 @@ HashDigest<SHA256> blockHash
...
+            Guid chainId,
+            Address address,
+            Block<T> lookupUntil)
+        {
+            if (lookupUntil is null)
+            {
+                throw new ArgumentNullException(nameof(lookupUntil));
+            }
+
+            if (_lastStateRefCaches.TryGetValue(chainId, out var stateRefCache)
```
Comment (longfin) [inline Libplanet/Store/DefaultStore.cs]:
```suggestion
            if (_lastStateRefCaches.TryGetValue(chainId, out LruCache<Address, Tuple<HashDigest<SHA256>, long>> stateRefCache)
```

File: Libplanet/Store/DefaultStore.cs [hunk 7]
```diff
@@ -564,6 +569,54 @@ HashDigest<SHA256> blockHash
...
+                && stateRefCache.TryGetValue(address, out var cache))
```
Comment (longfin) [inline Libplanet/Store/DefaultStore.cs]:
```suggestion
                && stateRefCache.TryGetValue(address, out Tuple<HashDigest<SHA256>, long>> cache))
```

File: Libplanet/Store/DefaultStore.cs [hunk 8]
```diff
@@ -619,6 +672,24 @@ HashDigest<SHA256> blockHash
...
+
+            if (!_lastStateRefCaches.ContainsKey(chainId))
+            {
+                _lastStateRefCaches[chainId] =
+                    new LruCache<Address, Tuple<HashDigest<SHA256>, long>>();
+            }
+
+            var stateRefCache = _lastStateRefCaches[chainId];
+
+            foreach (var address in addresses)
```
Comment (longfin) [inline Libplanet/Store/DefaultStore.cs]:
```suggestion
            foreach (Address address in addresses)
```

File: Libplanet/Store/IStore.cs [hunk 9]
```diff
@@ -224,6 +224,29 @@ void PutBlock<T>(Block<T> block)
...
+        /// <typeparam name="T">An <see cref="IAction"/> class used with
+        /// <paramref name="lookupUntil"/>.</typeparam>
+        /// <seealso cref="IStore.StoreStateReference(Guid, IImmutableSet{Address}, HashDigest{SHA256}, long)"/>
+        /// <seealso cref="IStore.IterateStateReferences(Guid, Address, long?, long?, int?)"/>
+#pragma warning restore MEN002
+        Tuple<HashDigest<SHA256>, long> LookupStateReference<T>(
+            Guid chainId,
+            Address address,
+            Block<T> lookupUntil)
+            where T : IAction, new();
```
Comment (dahlia) [inline Libplanet/Store/IStore.cs]: @earlbread Although this is too late, there still is `StoreExtension.LookupStateReference<T>(this IStore store, Guid chainId, Address address, Block<T> lookupUntil) where T : IAction, new()`, and indeed the extension method is still used here and there.  Isn't it intended, right?  Should I remove `StoreExtension.LookupStateReference<T>()` method?

Comment (earlbread) [reply Libplanet/Store/IStore.cs]: It was my mistake. 😨 I thought that I removed `StoreExtension` class, but I didn't. We should remove `StoreExtension.LookupStateReference<T>()`.

Comment (dahlia) [reply Libplanet/Store/IStore.cs]: Ok, nevermind. It was just for being sure. I am going to remove it in my next patch.

**Pull request merged**